### PR TITLE
[WIP] Proof of concept multi processing

### PR DESCRIPTION
**Pull request by piiswrong**
This is a proof of concept multi processing IO refactor. 
only works for linux/mac.

It can be tested with
```
import time
import numpy as np
import multiprocessing as mp
import mxnet as mx
from mxnet import gluon as gl


class ImageDataset(gl.data.Dataset):
	def __init__(self, N):
		self.N = N

	def __len__(self):
		return self.N
	
	def __getitem__(self, key):
		img = mx.image.imread('data/test_images/ILSVRC2012_val_00000001.JPEG')
		return mx.image.center_crop(img, (224, 224))[0]


if __name__ == '__main__':
	a = mx.nd.zeros((10,)) # trigger engine
	N = 10000
	M = 10
	BS = 100

	dataset = ImageDataset(N)
	data_loader = gl.data.DataLoader(dataset, BS, shuffle=True, num_workers=M)

	tic = time.time()

	for batch in data_loader:
		print(batch.shape)

	print(N/(time.time() - tic))
```
@winstywang @yajiedesign 

This also fixes the problem that mxnet doesn't work with fork() @andreaolgiati

### CODE REVIEW

File: include/mxnet/base.h [hunk 1]
```diff
@@ -200,7 +208,7 @@ struct Context {
     return true;
   }
   /*! \brief the maximal device type */
-  static const int32_t kMaxDevType = 4;
+  static const int32_t kMaxDevType = 6;
```
Comment (cjolivier01) [inline include/mxnet/base.h]: It's already here, but generally I see this sort of number derived from the last enum value, such as DevTypeMax or something.   Then you don't have to maintain it.

enum DeviceType {
      kCPU = cpu::kDevMask,
      kGPU = gpu::kDevMask,
      kCPUPinned = 3,
      kCPUShared = 5,
      kDeviceType /* or something like that */
    };

By the way, what happened to 4?

File: src/engine/threaded_engine_perdevice.cc [hunk 2]
```diff
@@ -50,6 +50,37 @@ class ThreadedEnginePerDevice : public ThreadedEngine {
...
+    gpu_normal_workers_.SignalForKill();
+    gpu_copy_workers_.SignalForKill();
+    cpu_normal_workers_.SignalForKill();
+    this->Stop();
+  }
+
+  void Stop() override {
+    SignalQueuesForKill();
+    gpu_normal_workers_.Clear();
+    gpu_copy_workers_.Clear();
```
Comment (cjolivier01) [inline src/engine/threaded_engine_perdevice.cc]: In LazyAllocArray, I suggest changing exit_now_ name to is_clearing_ or something like that, and then setting it back to false when the Clear() completes.
Before doing that, you probably can clear head_ and more_ , or at least set them to think they're empty and to start from the beginning again when filling.

File: include/mxnet/base.h [hunk 3]
```diff
@@ -143,7 +143,8 @@ struct Context {
   enum DeviceType {
     kCPU = cpu::kDevMask,
     kGPU = gpu::kDevMask,
-    kCPUPinned = 3
+    kCPUPinned = 3,
+    kCPUShared = 5,
```
Comment (cjolivier01) [inline include/mxnet/base.h]: Curious, why is this needed?

File: include/mxnet/base.h [hunk 4]
```diff
@@ -224,6 +232,12 @@ struct Context {
    */
   inline static Context CPUPinned(int32_t dev_id = -1);
   /*!
+   * Create a pinned CPU context.
```
Comment (cjolivier01) [inline include/mxnet/base.h]: Should this comment say something other than "pinned"?

File: include/mxnet/base.h (hunk 3 above)
Comment (piiswrong) [reply include/mxnet/base.h]: For creating shared memory from frontend

File: include/mxnet/base.h (hunk 1 above)
Comment (piiswrong) [reply include/mxnet/base.h]: Its a mask, so 101

**Comment by meanmee** [comment]
There is a bug when I run this script:
'module' object has no attribute 'set_start_method'

File: python/mxnet/gluon/data/dataloader.py [hunk 5]
```diff
@@ -87,10 +168,50 @@ def __init__(self, dataset, batch_size=None, shuffle=False, sampler=None,
...
+        if self._num_workers == 0:
+            for batch in self._batch_sampler:
+                yield self._batchify_fn([self._dataset[idx] for idx in batch])
+            return
+
+        key_queue = Queue()
+        data_queue = Queue(2*self._num_workers)
+
+        workers = []
+        for _ in range(self._num_workers):
```
Comment (ArtanisCV) [inline python/mxnet/gluon/data/dataloader.py]: Processes are created in each iteration. But it seems that the shared memory is not released when processes stop. So /dev/shm will exhaust quickly (full of "mx_pid_xxx") when num_workers > 1.

Comment (ArtanisCV) [inline python/mxnet/gluon/data/dataloader.py]: Is it possible to reuse shared memory in each iteration? Currently processes are re-created in each iteration, so shared memories are allocated & released a lot of times. On some older kernels (e.g., Linux 3.10 for Centos 7), shm_unlink will not free memory. So the program will OOM quickly when num_workers > 1.

Anyway, this is only related to compatibility. Just ignore it if it is not important.

**Comment by ArtanisCV** [comment]
> There is a bug when I run this script:
> 'module' object has no attribute 'set_start_method'

This is a new method in Py3.4, see https://docs.python.org/3/library/multiprocessing.html#multiprocessing.set_start_method

**Comment by ArtanisCV** [comment]
> There is a bug when I run this script:
> 'module' object has no attribute 'set_start_method'

@meanmee This is a new method in Py 3.4, see https://docs.python.org/3/library/multiprocessing.html#multiprocessing.set_start_method

File: python/mxnet/gluon/data/dataloader.py (hunk 5 above)
Comment (piiswrong) [reply python/mxnet/gluon/data/dataloader.py]: It can be reused. but needs to notify the original process. also memory needs to have the same size every time

**Comment by piiswrong** [comment]
moved to https://github.com/apache/incubator-mxnet/pull/8523
